source/tutorials/module-system/module-system.md

```diff
@@ -0,0 +1,1027 @@
++let
++
++  markerType = lib.types.submodule {
++    options = {
++      location = lib.mkOption {
++        type = lib.types.nullOr lib.types.str;
++        default = null;
++      };
++    };
++  };
++in {
++
++  options = {
++    map.markers = lib.mkOption {
++      type = lib.types.listOf markerType;
++    };
++  };
+ 
+ }
+```
+
+## Single option namespace: Defining markers in generate.requestParams
+
+Since all modules in `imports` are treated the same, we can also freely
+assign an option defined in our initial module. In this case we want to
+add some request parameters derived from the `map.markers` option, so
+that markers actually show up on the map.
+
+```diff
+diff --git a/marker.nix b/marker.nix
+index 6a6c686..c2c5da2 100644
+--- a/marker.nix
++++ b/marker.nix
+@@ -17,4 +17,26 @@ in {
+     };
+   };
+ 
++  config = {
++
++    map.markers = [
++      { location = "new york"; }
++    ];
++
++    generate.requestParams = let
++      paramForMarker = marker:
++        let
++          attributes =
++            [
++              "$(geocode ${
++                lib.escapeShellArg marker.location
++              })"
++            ];
++        in "markers=${
++          lib.concatStringsSep "\\|" attributes
++        }";
++    in map paramForMarker config.map.markers;
++
++  };
++
+ }
+```
+
+## Set map.{center,zoom} for more than {1,2} markers
+
+Let's let the API infer the center if we have more than one marker, and
+let's let it infer the zoom as well if there's more than 2.
+
+```diff
+diff --git a/marker.nix b/marker.nix
+index c2c5da2..b80ddd0 100644
+--- a/marker.nix
++++ b/marker.nix
+@@ -23,6 +23,14 @@ in {
+       { location = "new york"; }
+     ];
+ 
++    map.center = lib.mkIf
++      (lib.length config.map.markers >= 1)
++      null;
++
++    map.zoom = lib.mkIf
++      (lib.length config.map.markers >= 2)
++      null;
++
+     generate.requestParams = let
+       paramForMarker = marker:
+         let
+```
+
+## Nested submodules: Introducing users option
+
+We will now introduce the option `users`, where we make use of a very
+useful type, `lib.types.attrsOf <subtype>`. This type lets us specify an
+attribute set as a value, where the keys can be arbitrary, but each
+value has to conform to the given <subtype>.
+
+In this case, we'll use another submodule as the subtype, one that
+allows declaring a departure marker, which notably also makes use of our
+`markerType` submodule, giving us a nested structure of submodules.
+
+We're now propagating each of the users marker definitions to the
+`map.markers` option.
+
+```diff
+diff --git a/marker.nix b/marker.nix
+index b80ddd0..3c54ad8 100644
+--- a/marker.nix
++++ b/marker.nix
+@@ -9,9 +9,24 @@ let
+       };
+     };
+   };
++
++  userType = lib.types.submodule {
++    options = {
++      departure = lib.mkOption {
++        type = markerType;
++        default = {};
++      };
++    };
++  };
++
+ in {
+ 
+   options = {
++
++    users = lib.mkOption {
```

Comment:
This is as far as I got before I mentally ran out of gas.

Reply:
All of Nix is running on fumes, having gas is clearly a luxury.